### DISCUSSION TITLE
The `constraintsFiltering` only works on direct dependencies

### DISCUSSION THREAD

**Discussion by han3zeng** [renovatebot/renovate#26954] [Request Help] [answered]
### How are you running Renovate?

Mend Renovate hosted app on github.com

### If you're self-hosting Renovate, select which platform you are using.

None

### What is your question?

I have a repository that has package.json containing `{ "engines": { "node": "^16" } }` setting. Even though I've set `constraintsFiltering` to `strict`, Renovate still opens a PR intending to update a package, including a dependency with package.json specifying `{ "engines": { "node": ">=18" } }`.

Is it correct that the `"constraintsFiltering": "strict"` only works on direct dependency?  If so, is there a common practice to filter out the package that has either direct or indirect dependencies that do not conform with constraint setting of my dependent project?

Here is the [repository](https://github.com/han3zeng/renovate-trial/pull/3) to demonstrate the case. Renovate opens a PR updating `nested-obsolete-dependencies` from version 1.0.4 to version 1.0.6. The version 1.0.6 has a dependency on @luxass/strip-json-comments, which specifies `{ "engines": { "node": ">=18" } }`.

Thanks for your time and help.

### Logs (if relevant)

<details><summary>Logs</summary>

```

Copy/paste the relevant log(s) here, between the starting and ending backticks

```

</details>

**Comment by rarkins** [comment, marked as answer]
This is intentional behavior by Renovate. Looking at your reproduction:

- The project declares engines=^16
- It depends on "nested-obsolete-dependencies": "^1.0.4"
- This is resolved to 1.0.4 in the lock file, which also has engines=^16 so that's fine
- It's updated to 1.0.6 which again has engines=^16 which is fine
- But 1.0.6 has added @luxass/strip-json-comments	"^1.2.0" which resolves to 1.2.0 which has engines >=18

This means that the mistake/problem was made in nested-obsolete-dependencies@1.0.6 which added an incompatible dependency.

Renovate "trusts" the declarations of packages so if nested-obsolete-dependencies says it supports node ^16 then we trust that. If it has a mistake then it needs to be caught during the "npm install". I'm not sure if it has a "strict engines check" capability these days or not

**Reply by viceice** [reply]
- https://docs.npmjs.com/cli/v8/using-npm/config#engine-strict
